File: src/main/java/com/icthh/xm/ViewServer.java
package com.icthh.xm;

import com.icthh.xm.utils.SocketUtils;
import org.apache.commons.io.IOUtils;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.DefaultServlet;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.eclipse.jetty.util.resource.Resource;

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Logger;

import static java.nio.charset.StandardCharsets.UTF_8;

public class ViewServer {

    private static final Logger LOGGER = java.util.logging.Logger.getLogger(ViewServer.class.getSimpleName());

    public static Server embeddedServer;
    private static Integer serverPort = 54321;
    private static AtomicBoolean isInited = new AtomicBoolean(false);
    public static boolean isDev = false;// Boolean.parseBoolean(System.getenv("IS_DEV_PLUGIN_RUN"));

    public static String getServerUrl() {
        if (isDev) {
            return "http://localhost:4200";
        }
        return "http://localhost:" + serverPort;
    }

    public static void startServer() throws Exception {
        if (isInited.compareAndSet(false, true)) {

            URL url = ViewServer.class.getClassLoader().getResource("/static/");
            URI webRootUri = url.toURI();
            ServletContextHandler context = new ServletContextHandler(
                    ServletContextHandler.SESSIONS);
            context.setContextPath("/");
            context.setBaseResource(Resource.newResource(webRootUri));
            context.setWelcomeFiles(new String[] { "index.html" });

            ServletHolder holderPwd = new ServletHolder("default", DefaultServlet.class);
            holderPwd.setInitParameter("dirAllowed", "true");
            context.addServlet(holderPwd, "/*");
            context.addServlet(AppServlet.class, "/app");

            serverPort = SocketUtils.Companion.findAvailableTcpPort();

            Server embeddedServer = new Server(serverPort);
            embeddedServer.setHandler(context);
            embeddedServer.start();
            embeddedServer.dump(System.err);
            ViewServer.embeddedServer = embeddedServer;

            LOGGER.info("\n\n\nServer url:" + getServerUrl() + "\n\n\n");
        }
    }

    public static class AppServlet extends HttpServlet {
        private static final long serialVersionUID = 1L;
        private static final String index = readIndexHtml();

        public AppServlet() {
        }

        private static String readIndexHtml() {
            try {
                return IOUtils.toString(AppServlet.class.getClassLoader().getResourceAsStream("/static/index.html"), UTF_8);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
            response.setContentType("text/html;charset=UTF-8");
            var out = response.getOutputStream();
            out.print(index.replace("${pipeId}", request.getQueryString()));
        }
    }

}

File: src/main/kotlin/com/icthh/xm/xmeplugin/ViewServer.kt
package com.icthh.xm.xmeplugin

import com.icthh.xm.xmeplugin.utils.SocketUtils
import com.icthh.xm.xmeplugin.utils.log
import org.apache.commons.io.IOUtils
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.servlet.DefaultServlet
import org.eclipse.jetty.servlet.ServletContextHandler
import org.eclipse.jetty.servlet.ServletHolder
import org.eclipse.jetty.util.resource.Resource
import java.io.IOException
import java.nio.charset.StandardCharsets
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * This class is responsible for starting the server that will serve the web view dialogs like settings.
 */
object ViewServer {

    private var embeddedServer = lazy { createServer() }

    @Volatile
    var serverPort = SocketUtils.findAvailableTcpPort()
    var isDev: Boolean = false // Boolean.parseBoolean(System.getenv("IS_DEV_PLUGIN_RUN"));

    val serverUrl: String
        get() {
            if (isDev) {
                return "http://localhost:4200"
            }
            return "http://localhost:$serverPort"
        }

    @Throws(Exception::class)
    fun startServer() {
        if (!embeddedServer.isInitialized()) {
            embeddedServer.value
            log.info("\n\n\n\n\nServer url:$serverUrl\n\n\n\n\n")
        }
    }

    private fun createServer(): Server {
        val context = ServletContextHandler(ServletContextHandler.SESSIONS)

        var url = AppServlet::class.java.classLoader.getResource("static")
        if (url == null) {
            url = AppServlet::class.java.classLoader.getResource("static/")
            context.welcomeFiles = arrayOf("index.html")
        } else {
            context.welcomeFiles = arrayOf("/index.html")
        }
        val webRootUri = url?.toURI()
        context.setContextPath("/")
        context.setBaseResource(Resource.newResource(webRootUri))

        val holderPwd = ServletHolder("default", DefaultServlet::class.java)
        holderPwd.setInitParameter("dirAllowed", "true")
        context.addServlet(holderPwd, "/*")
        context.addServlet(AppServlet::class.java, "/app")

        val embeddedServer = Server(serverPort)
        embeddedServer.setHandler(context)
        embeddedServer.dump(System.err)
        embeddedServer.start()
        return embeddedServer
    }

    class AppServlet : HttpServlet() {

        override fun doGet(request: HttpServletRequest, response: HttpServletResponse) {
            response.contentType = "text/html;charset=UTF-8"
            val out = response.outputStream
            out.print(index.replace("\${pipeId}", request.queryString))
        }

        companion object {
            private const val serialVersionUID = 1L
            private val index = readIndexHtml()

            private fun readIndexHtml(): String {
                try {
                    AppServlet::class.java.classLoader.getResourceAsStream("static/index.html").use { `is` ->
                        return IOUtils.toString(`is`, StandardCharsets.UTF_8)
                    }
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            }
        }
    }
}

